app/src/main/java/com/github/geohunt/app/model/database/firebase/FirebaseDatabase.kt

```diff
@@ -297,6 +300,69 @@ class FirebaseDatabase(activity: Activity) : Database {
     override suspend fun unfollow(follower: String, followee: String) {
         doFollow(follower, followee, follow = false)
     }
-}
 
+    /**
+     * Returns a list of users that liked a given challenge
+     * @param cid the challenge id
+     * @return a list of users that liked the challenge
+     */
+    internal fun getLikeRefById(cid: String): LazyRef<User> {
+        return object : BaseLazyRef<User>() {
+            override fun fetchValue(): Task<User> {
+                return dbChallengeRef.child(cid).child("likes").get().thenMap {
+                    val uid = it.key!!
+                    getUserById(uid).fetch().result
+                }
+            }
+
+            override val id: String
+                get() = cid
+        }
+    }
+
+    /**
+     * Insert a like for a user into the Firebase
+     * @param uid the user id
+     * @param cid the challenge id
+     * @return a task that will complete when the like is inserted
+     */
+    override fun insertUserLike(uid: String, cid: String): Task<Void> {
+        //Add the challenge to the user's liked challenges
+        return Tasks.whenAll(
+            dbUserRef.child(uid).child("likes").child(cid).setValue(true),
+            dbChallengeRef.child(cid).child("likes").child(uid).setValue(true)
+        )
+    }
+
+    /**
+     * Remove a like for a user from the Firebase
+     * @param uid the user id
+     * @param cid the challenge id
+     * @return a task that will complete when the like is removed
+     */
+    override fun removeUserLike(uid: String, cid: String): Task<Void> {
+        //Remove the challenge from the user's liked challenges
+        return Tasks.whenAll(
+            dbChallengeRef.child(cid).child("likes").child(uid).removeValue(),
+            dbUserRef.child(uid).child("likes").child(cid).removeValue()
+        )
```

Comment:
There was another implementation for the number of like, but it is not important now because I have changed the implementation of the code following your review. The new functionality ensures that a user cannot like twice a challenge or unlike a challenge that he or she does not like.